File: cdmsbats/HistSubRecord.py
# -*- coding: utf-8 -*-
"""
Created on Fri Apr 11 12:27:34 2014

@author: tdoughty1
"""

from numpy import copy

from DataRecord import DataRecord


class HistSubRecord(DataRecord):

    def _InitValues(self):
        self._Masks = []
        self._Times = None

    def _PrintLine(self):
        print "Found History Buffer Sub Record."

    # Primary Function Call for data record
    def ReadRecord(self, Record, index, debug=False):

        # Get number of times stored
        nTimes = Record[index]
        index += 1

        # Store number of times
        start = index + 1
        end = start + nTimes
        self._Times = copy(Record[start:end])
        index += nTimes

        # Get number of masks stored
        nMasks = Record[index]
        index += 1

        # Store all masks for all times
        for i in range(nTimes):
            start = index
            end = start + nMasks
            self._Masks.append(copy(Record[start:end]).astype('uint32'))
            index += nMasks

        if debug:
            self.PrintValues()

        return index

    def StoreValues(self, Record):
        pass

    def PrintValues(self):

        print len(self._Masks)

        nTime = 0
        for time in self._Times:
            nMask = 0
            nTime += 1
            print "time %d = %d" % (nTime, time)
            for mask in self._Masks[nTime-1]:
                nMask += 1
                print "time %d - mask %d = 0x%x" % (nTime, nMask, mask)

File: cdmsbats/histsubrecord.cpp
#include <iostream>
#include <iomanip>

#include "histsubrecord.h"

using namespace std;

// Primary Function Call for data record
void HistSubRecord::ReadRecord(CDMSRawFileStream* filePtr, bool debug=false)
{
    // Get number of times stored
    int32_t nT[1];
    filePtr->ReadWords(4, nT);
    _nTimes = nT[0];

    /////////////////////////////////////////////////////
    // Store times
    /////////////////////////////////////////////////////
    int32_t tempTArray[_nTimes];
    filePtr->ReadWords(_nTimes*4, tempTArray);
    _Times.resize(_nTimes);

    for(int i=0; i<_nTimes; i++)
    {
        _Times.at(i) = tempTArray[i];
    }

    // Get number of masks stored
    int32_t nM[1];
    filePtr->ReadWords(4, nM);
    _nMasks = nM[0];

    /////////////////////////////////////////////////////
    // Store all masks for each time
    /////////////////////////////////////////////////////

    // Define temporary array and vector for convenience
    uint32_t tempMArray[_nMasks];
    vector<uint32_t> tempVec;

    for(int i=0; i < _nTimes; i++)
    {
        // Get array of Masks
        filePtr->ReadWords(_nMasks*4, tempMArray);

        // Store Masks in a vector
        for(int j=0; j < _nMasks; j++)
        {
            tempVec.push_back(tempMArray[j]);
        }
        _Masks.push_back(tempVec);
        tempVec.clear();
    }
}

void HistSubRecord::PrintValues()
{
    for(int i=0; i < _nTimes; i++)
    {
        cout << "time " << setbase(10) << (i+1) << " = " << _Times[i] << endl;

        for(int j=0; j < _nMasks; j++)
        {
            cout << "time " << setbase(10) << (i+1) << " - mask " << (j+1);
            cout << " = 0x" << setbase(16) << _Masks[i][j] << endl;
        }
    }
}
